tests/test_bootloader.py

```diff
@@ -0,0 +1,634 @@
+    port1 : string
+      Filepath to the COM port of the Piksi Under Test.
+    port2 : string
+      Filepath to the COM port of a Piksi connected via UART to the
+      Piksi Under Test.
+    stm_fw : intelhex.IntelHex
+      STM firmware to use in tests.
+    nap_fw : intelhex.IntelHex
+      NAP firmware to use in tests.
+    verbose : bool
+      Print status output.
+    n_queue : int
+      Number of flash operations to queue.
+    skip_single : bool
+      Skip single Piksi tests.
+    skip_double : bool
+      Skip double Piksi tests.
+    """
+    super(TestBootloader, self).__init__(testname)
+    self.port1 = port1
+    self.port2 = port2
+    self.stm_fw = stm_fw
+    self.nap_fw = nap_fw
+    self.verbose = verbose
+    self.n_queue = n_queue
+    self.skip_single = skip_single
+    self.skip_double = skip_double
+
+  def tearDown(self):
+    """ Clean up after running each test. """
+    # Hack: Wait a bit for 'Piksi Disconnected'
+    # print from sbp.client.handler.Handler
+    time.sleep(0.1)
+
+  def _piksi_settings_cb(self):
+    """ Callback to set flag indicating Piksi settings have been received. """
+    self.settings_received = True
+
+  def get_piksi_settings(self, handler):
+    """
+    Get settings from Piksi.
+
+    Parameters
+    ==========
+    handler : sbp.client.handler.Handler
+      Handler to send messages to Piksi over and register callbacks with.
+    """
+
+    self.settings_received = False
+
+    with SettingsView(handler,
+                      read_finished_functions=[self._piksi_settings_cb],
+                      gui_mode=False) as sv:
+      with Timeout(TIMEOUT_READ_SETTINGS) as timeout:
+        sv._settings_read_button_fired()
+        while not self.settings_received:
+          time.sleep(0.1)
+
+    return sv.settings
+
+  def test_get_versions(self):
+    """ Get Piksi bootloader/firmware/NAP version from device. """
+    unittest.skipIf(self.skip_single, 'Skipping single Piksi tests')
+
+    if self.verbose: print "--- test_get_versions ---"
+
+    with serial_link.get_driver(use_ftdi=False, port=self.port1) as driver:
+      with Handler(driver.read, driver.write) as handler:
+
+        with Bootloader(handler) as piksi_bootloader:
+
+          # Get bootloader version, print, and jump to application firmware.
+          with Timeout(TIMEOUT_BOOT) as timeout:
+            if self.verbose: print "Waiting for bootloader handshake"
+            piksi_bootloader.handshake()
+          piksi_bootloader.jump_to_app()
+          print "Piksi Bootloader Version:", piksi_bootloader.version
+
+        # Wait for heartbeat, get settings, print firmware/NAP versions.
+        heartbeat = Heartbeat()
+        handler.add_callback(heartbeat, SBP_MSG_HEARTBEAT)
+        if self.verbose: print "Waiting to receive heartbeat"
+        while not heartbeat.received:
+          time.sleep(0.1)
+        if self.verbose: print "Received hearbeat"
+        handler.remove_callback(heartbeat, SBP_MSG_HEARTBEAT)
```

Comment:
Not sure I follow how that can be done - doesn't look like `handler.wait` can be passed a callback or something?